### PR TITLE
Do not add visual content if activity is finishing (fixes #312).

### LINKED ISSUE #312: BadTokenException in RestAPI
```
android.view.WindowManager$BadTokenException: Unable to add window -- token android.os.BinderProxy@43237e58 is not valid; is your activity running?
at android.view.ViewRootImpl.setView(ViewRootImpl.java:536)
at android.view.WindowManagerGlobal.addView(WindowManagerGlobal.java:259)
at android.view.WindowManagerImpl.addView(WindowManagerImpl.java:69)
at android.app.Dialog.show(Dialog.java:294)
at android.app.AlertDialog$Builder.show(AlertDialog.java:966)
at com.nutomic.syncthingandroid.syncthing.RestApi.requireRestart(RestApi.java:369)
at com.nutomic.syncthingandroid.syncthing.RestApi$10.onDeviceIdNormalized(RestApi.java:744)
at com.nutomic.syncthingandroid.syncthing.RestApi$11.onPostExecute(RestApi.java:889)
at com.nutomic.syncthingandroid.syncthing.RestApi$11.onPostExecute(RestApi.java:876)
at android.os.AsyncTask.finish(AsyncTask.java:632)
at android.os.AsyncTask.access$600(AsyncTask.java:177)
at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:645)
at android.os.Handler.dispatchMessage(Handler.java:102)
at android.os.Looper.loop(Looper.java:212)
at android.app.ActivityThread.main(ActivityThread.java:5135)
at java.lang.reflect.Method.invokeNative(Native Method)
at java.lang.reflect.Method.invoke(Method.java:515)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:878)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)
at dalvik.system.NativeStart.main(Native Method)
```

### PR DESCRIPTION
**Pull request by Zillode**
- I did a check on all fragments for use-while-finishing usage patterns
- If it is finishing and we need to restart, we just restart.

### CODE REVIEW

File: src/main/java/com/nutomic/syncthingandroid/fragments/DevicesFragment.java [hunk 1]
```diff
@@ -54,7 +54,7 @@ public void onActivityCreated(Bundle savedInstanceState) {
 
     private void initAdapter() {
         SyncthingActivity activity = (SyncthingActivity) getActivity();
-        if (activity == null || activity.getApi() == null)
+        if (activity == null || activity.getApi() == null || activity.isFinishing())
```
Comment (Nutomic) [inline src/main/java/com/nutomic/syncthingandroid/fragments/DevicesFragment.java]: Just do `getActivity().isFinishing()`.

**Comment by Nutomic** [comment]
Are those all places from the logs?

Btw it's also in `MainActivity.onApiChange()`, even though that has `isFinishing()`.

File: src/main/java/com/nutomic/syncthingandroid/fragments/DevicesFragment.java (hunk 1 above)
Comment (Zillode) [inline src/main/java/com/nutomic/syncthingandroid/fragments/DevicesFragment.java]: that would look wierd, we have the activity reference just the line above it and test its validity in the same if test

**Comment by Zillode** [comment]
It is?:/
Where do you see those logs?

File: src/main/java/com/nutomic/syncthingandroid/fragments/DevicesFragment.java (hunk 1 above)
Comment (Nutomic) [inline src/main/java/com/nutomic/syncthingandroid/fragments/DevicesFragment.java]: Ah sorry misread, I thought you added that cast.

**Comment by Nutomic** [comment]
Eg this:
```
java.lang.RuntimeException: Unable to start service com.nutomic.syncthingandroid.syncthing.SyncthingService@41d2eea0 with Intent { cmp=com.nutomic.syncthingandroid/.syncthing.SyncthingService (has extras) }: android.view.WindowManager$BadTokenException: Unable to add window -- token android.os.BinderProxy@41d14348 is not valid; is your activity running?
at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:2743)
at android.app.ActivityThread.access$2100(ActivityThread.java:144)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1302)
at android.os.Handler.dispatchMessage(Handler.java:102)
at android.os.Looper.loop(Looper.java:136)
at android.app.ActivityThread.main(ActivityThread.java:5146)
at java.lang.reflect.Method.invokeNative(Native Method)
at java.lang.reflect.Method.invoke(Method.java:515)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:732)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:566)
at de.robv.android.xposed.XposedBridge.main(XposedBridge.java:132)
at dalvik.system.NativeStart.main(Native Method)
Caused by: android.view.WindowManager$BadTokenException: Unable to add window -- token android.os.BinderProxy@41d14348 is not valid; is your activity running?
at android.view.ViewRootImpl.setView(ViewRootImpl.java:532)
at android.view.WindowManagerGlobal.addView(WindowManagerGlobal.java:259)
at android.view.WindowManagerImpl.addView(WindowManagerImpl.java:69)
at android.app.Dialog.show(Dialog.java:286)
at android.app.AlertDialog$Builder.show(AlertDialog.java:951)
at com.nutomic.syncthingandroid.syncthing.SyncthingService.showDisabledDialog(SyncthingService.java:418)
at com.nutomic.syncthingandroid.activities.MainActivity.onApiChange(MainActivity.java:57)
at com.nutomic.syncthingandroid.syncthing.SyncthingService.onApiChange(SyncthingService.java:384)
at com.nutomic.syncthingandroid.syncthing.SyncthingService.updateState(SyncthingService.java:214)
at com.nutomic.syncthingandroid.syncthing.SyncthingService.onStartCommand(SyncthingService.java:139)
at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:2726)
... 11 more
```

**Comment by Zillode** [comment]
I believe [this](https://github.com/Zillode/syncthing-android/commit/4e75d18c020f656faad05a157bf02bb819feec0b) is the proper way to do the dialogs. Their construction will be done on the GUI thread so we know the activity will not finish during this process.

**Comment by Nutomic** [comment]
`Activity#runOnUiThread()` seems better for this. Though it doesn't seem a threading problem.

**Comment by Zillode** [comment]
No because you can't return values from runOnUiThread to be able to cancel the dialog.
It is a threading problem because onApiChange happens outside the UI thread, so creating an alert there is not correct

**Comment by Nutomic** [comment]
But you don't need to return anything, you can use a class member (like it is now). Also, your code won't even work because [AsyncTask has to be constructed on the GUI thread](http://developer.android.com/reference/android/os/AsyncTask.html#AsyncTask%28%29).

So like I said, putting `runOnUiThread()` in `onApiChanged()` should have the same effect.

**Pull request closed without merging**